File: src/main/java/org/stt/config/ConfigServiceFacade.java
package org.stt.config;

import org.stt.Service;

import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton
public class ConfigServiceFacade implements ConfigService, Service {
    private final YamlConfigService yamlConfigService;
    private final JsonConfigService jsonConfigService;
    private boolean skipYaml;

    @Inject
    public ConfigServiceFacade(YamlConfigService yamlConfigService, JsonConfigService jsonConfigService) {
        this.yamlConfigService = yamlConfigService;
        this.jsonConfigService = jsonConfigService;
    }

    @Override
    public void start() throws Exception {
        jsonConfigService.start();
        if (jsonConfigService.isNewConfig()) {
            yamlConfigService.start();
        } else {
            skipYaml = true;
        }
    }

    @Override
    public void stop() {
        jsonConfigService.stop();
        if (!skipYaml) {
            yamlConfigService.stop();
        }
    }

    @Override
    public ConfigRoot getConfig() {
        return skipYaml ? jsonConfigService.getConfig() : yamlConfigService.getConfig();
    }
}

File: src/main/kotlin/org/stt/config/ConfigServiceFacade.kt
package org.stt.config

import org.stt.Service

import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class ConfigServiceFacade @Inject
constructor(private val yamlConfigService: YamlConfigService, private val jsonConfigService: JsonConfigService) : ConfigService, Service {
    private var skipYaml: Boolean = false

    override val config: ConfigRoot
        get() = if (skipYaml) jsonConfigService.config else yamlConfigService.config

    override fun start() {
        jsonConfigService.start()
        if (jsonConfigService.isNewConfig) {
            yamlConfigService.start()
            jsonConfigService.config = yamlConfigService.config
        } else {
            skipYaml = true
        }
    }

    override fun stop() {
        jsonConfigService.stop()
        if (!skipYaml) {
            yamlConfigService.stop()
        }
    }
}
